### ISSUE TITLE
There's no way to exit a session on wf-shell

### ISSUE THREAD

**Issue by ericonr** [WayfireWM/wf-shell#41]
There was no easy way to exit a session through the wf-shell, I had to open a terminal and run `kill wayfire` to do so. It's not much work to do it this way, but I believe it could be implemented and make for a better user experience.

Perhaps even create a "canonical" way of exiting a session that doesn't rely on simply killing a process. I'm open to helping with developing this, because this is a project I'm really interested in right now!

**Comment by soreau** [comment]
To exit wayfire, you can use Ctrl+Alt+Backspace.

There is not a way to do this with wf-shell yet as far as I know.

**Comment by ericonr** [comment]
> To exit wayfire, you can use Ctrl+Alt+Backspace.

Is this documented anywhere? I couldn't find it.

Even so, I believe it would be a good idea if wf-shell could be used to exit a session (or even turn off a system), because it seems like it's a system designed in a way that can be used primarily by mouse as well.

It's open to discussion how this might be best implemented.

**Comment by soreau** [comment]
> > To exit wayfire, you can use Ctrl+Alt+Backspace.
> 
> Is this documented anywhere? I couldn't find it.

I don't think so. I also couldn't find any documentation on how to start wayfire. Perhaps it could be added somewhere like the wiki.

> Even so, I believe it would be a good idea if wf-shell could be used to exit a session (or even turn off a system), because it seems like it's a system designed in a way that can be used primarily by mouse as well.
> 
> It's open to discussion how this might be best implemented.

Perhaps we could have string options that point to scripts for logout, reboot and shutdown. If the option is set, a button is created in the menu for it. By default the options would not be set so the buttons wouldn't exist but we could add some example scripts to the wiki and when used, the buttons would automatically show and execute the respective script. @ammen99 what do you think?

**Comment by ammen99** [comment]
The biggest problem with such a feature is that if you have multiple Wayfire instances, you want to only close the current one. So Wayfire and wf-shell would need to communicate via either DBus or via a Wayland protocol. I can imagine that we can have a small wayland protocol plugin in Wayfire which accepts such commands, and a small widget in wf-shell which uses this protocol.

**Comment by soreau** [comment]
@ammen99 True indeed, I had considered a protocol first but thought this might be too complex. Do we have any example of a plugin implementing a protocol yet?

**Comment by ammen99** [comment]
No, but it is not hard to do. Just create a singleton plugin like autostart, and in the constructor set up the wl_global and callbacks.

**Comment by soreau** [comment]
@ericonr If you feel like this is something you'd like to do, you probably should join #wayfire on irc.freenode.net where we can chat. I must say that for a newcomer, this task might be a bit daunting, though we should be able to help you through it.

**Comment by ericonr** [comment]
Yeah, the task seems to have increased in scope a bit. I felt the IPC interface on sway, using a socket, works pretty well overall. Is there any reason in particular this project doesn't use that?

Is making a wayland protocol exclusively for wayfire logout a "sustainable" endeavor? I don't know much about this, so I could be talking nonsense, though. Any reason you thought about implementing a protocol instead of a DBus interface?

I just joined your channel on freenode as well!

**Comment by soreau** [comment]
> Is making a wayland protocol exclusively for wayfire logout a "sustainable" endeavor? I don't know much about this, so I could be talking nonsense, though. Any reason you thought about implementing a protocol instead of a DBus interface?

As @ammen99 outlined, there must be some sort of IPC, whether it be DBus, protocol or another socket. I assume wayland protocol by default since it's native and wayfire doesn't currently use DBus or any other IPC.

**Comment by soreau** [comment]
https://github.com/WayfireWM/wf-shell/issues/30 offers an option, which I've added to the wiki.

**Comment by ammen99** [comment]
Actually if you don't run multiple Wayfire sessions on the same machine, you can also just add a launcher which does `pkill wayfire`

**Comment by nwg-piotr** [comment]
I first installed Wayfire last night, and got to this thread in search of solution to [this issue](https://github.com/nwg-piotr/nwg-launchers/issues/108). While answering, I did not yet know that there is no command to exit the session. It would be really useful. @ericonr , if you know a possible workaround, please join the thread.

**Comment by soreau** [comment]
There is no integrated logout button in wf-shell but there is Ctrl+Alt+Backspace to end the wayfire session or you can look [here](https://github.com/WayfireWM/wayfire/wiki/Tutorial#logout-and-shutdown) for wlogout.

**Comment by nwg-piotr** [comment]
> or you can look here for wlogout

I know, but this is not a solution to this certain issue. To have a "Logout" button in `nwgbar`, a command is necessary, not another set of buttons.

**Comment by epsilon-0** [comment]
@soreau isn't wf-shell just an application that is started by wayfire? So exiting wf-shell shouldn't be the goal (I think).
From the minimal look over `nwgbar`, it seems to be using `swaymsg` to send a command (or is it an api call??) to sway to logout.
Something like an invoker for wayfire api would be the desirable outcome, @nwg-piotr correct me if I am wrong.
I guess the dbus thing mentioned above is aiming to solve this.

I defer judgements and decisions.

**Comment by soreau** [comment]
> @soreau isn't wf-shell just an application that is started by wayfire? So exiting wf-shell shouldn't be the goal (I think).

I don't understand what you mean by this. Why would you only want to shutdown the shell apps? I thought the goal was to run a command that either a) exits wayfire immediately b) start a gui app that allows you to choose shutdown/logout/etc. I thought wlogout provides b.

> From the minimal look over `nwgbar`, it seems to be using `swaymsg` to send a command (or is it an api call??) to sway to logout.

Can't it just call wlogout instead of swaymsg or whatever?

> Something like an invoker for wayfire api would be the desirable outcome, @nwg-piotr correct me if I am wrong.
> I guess the dbus thing mentioned above is aiming to solve this.
> 
> I defer judgements and decisions.

**Comment by epsilon-0** [comment]
> > @soreau isn't wf-shell just an application that is started by wayfire? So exiting wf-shell shouldn't be the goal (I think).
> 
> I don't understand what you mean by this. Why would you only want to shutdown the shell apps? I thought the goal was to run a command that either a) exits wayfire immediately, or b) start a gui app that allows you to choose shutdown/logout/etc. I thought wlogout provides b.
> 

I am actually agreeing with what you said, I don't know if I phrased it weirdly.
OTOH, I don't think even wlogout provides (b) as even wlogout needs the logout button to be mapped to a command (though I haven't tested this theory out).

> > From the minimal look over `nwgbar`, it seems to be using `swaymsg` to send a command (or is it an api call??) to sway to logout.
> 
> Can't it just call wlogout instead of swaymsg or whatever?

What the result would be 
 Open `nwgbar` > Press `logout` button > Opens `wlogout` > Press `logout` button (which does nothing, as there's still no logout command)

Both `wlogout` and `nwgbar` need some command to map the buttons to.

> 
> > Something like an invoker for wayfire api would be the desirable outcome, @nwg-piotr correct me if I am wrong.
> > I guess the dbus thing mentioned above is aiming to solve this.
> > I defer judgements and decisions.

**Comment by wb9688** [comment]
If using `logind`, you could use `loginctl terminate-user $USER`, which is also set by default as logout command in `wlogout`.

**Comment by soreau** [comment]
> What the result would be
> Open `nwgbar` > Press `logout` button > Opens `wlogout` > Press `logout` button (which does nothing, as there's still no logout command)


Ah, I didn't realize wlogout doesn't actually do anything.

**Comment by soreau** [comment]
> If using `logind`, you could use `loginctl terminate-user $USER`, which is also set by default as logout command in `wlogout`.

Then again, maybe it does?

**Comment by epsilon-0** [comment]
@soreau interesting workaround, then this should work similarly for nwgbar, both are basically similar for end user command configuration.
Though this still has the same aforementioned issues 
> As @ammen99 outlined, there must be some sort of IPC, whether it be DBus, protocol or another socket. I assume wayland protocol by default since it's native and wayfire doesn't currently use DBus or any other IPC.

**Comment by soreau** [comment]
And there *is* a [dbus plugin](https://github.com/damianatorrpm/wayfire-plugin_dbus_interface) now, though it isn't upstreamed to wayfire. Maybe wayfire-plugins-extra could house it but I'm not sure how @ammen99 would feel about this..

**Comment by nwg-piotr** [comment]
> you could use `loginctl terminate-user $USER`

On my machine this ends the wayfire / sway session, but it hangs afterwards.

Finally - nwgbar is just a button bar, nothing else. It's up to the user which commands to use. As far as I know, the same problem appears on dwm.

**Comment by ammen99** [comment]
@nwg-piotr I just asked how a compositor should be exited from a panel or whatever, and I was told that the following can be done:

1. Open the wayland socket which is typically `$XDG_RUNTIME_DIR/$WAYLAND_DISPLAY`
2. Use SO_PEERCRED to get the PID of the compositor on that socket
3. Send SIGINT to the compositor

All of this can happen in a small C program, or maybe a python script.

On a side note, wayfire doesn't handle SIGINT properly, but that is a problem with Wayfire itself.

**Comment by nwg-piotr** [comment]
One way or another, the [nwgbar](https://github.com/nwg-piotr/nwg-launchers#nwgbar) command, a part of [nwg-launchers](https://github.com/nwg-piotr/nwg-launchers), is just a configurable button bar, which does nothing else but displays a horizontal or vertical row of buttons assigned to user-defined commands. The sample template, which comes preinstalled, creates the exit menu for sway on Arch Linux:

```json
[
  {
    "name": "Lock screen",
    "exec": "swaylock -f -c 000000",
    "icon": "system-lock-screen"
  },
  {
    "name": "Logout",
    "exec": "swaymsg exit",
    "icon": "system-log-out"
  },
  {
    "name": "Reboot",
    "exec": "systemctl reboot",
    "icon": "system-reboot"
  },
  {
    "name": "Shutdown",
    "exec": "systemctl -i poweroff",
    "icon": "system-shutdown"
  }
]
```

However, this may be ANY set of buttons with ANY commands. Whatever could be called here to exit Wayfire, must be an external command/script.

**Comment by ammen99** [comment]
I understand, that's why I said the above steps can be executed by a small script which can then be put as a command there. I will try to put something together in the next few days and see how well it works.

**Comment by nwg-piotr** [comment]
Thanks in advance! Also on behalf of @epsilon-0 , I believe. :)

**Comment by soreau** [comment]
> @nwg-piotr I just asked how a compositor should be exited from a panel or whatever, and I was told that the following can be done:
> 
>     1. Open the wayland socket which is typically `$XDG_RUNTIME_DIR/$WAYLAND_DISPLAY`
> 
>     2. Use SO_PEERCRED to get the PID of the compositor on that socket
> 
>     3. Send SIGINT to the compositor
> 
> 
> All of this can happen in a small C program, or maybe a python script.
> 
> On a side note, wayfire doesn't handle SIGINT properly, but that is a problem with Wayfire itself.

I have created https://github.com/soreau/wayland-logout based on this.

**Comment by nwg-piotr** [comment]
> I have created https://github.com/soreau/wayland-logout based on this.

It would be great to make it available to users in some way.

At least I'll give it a try. Tomorrow, as it's almost 4 am in my place. I was working on displays detection in [Azote](https://github.com/nwg-piotr/azote) for it to work on Wayfire. Seems to be all right now (not yet merged).

**Comment by soreau** [comment]
Also I have pushed a [python version](https://github.com/soreau/wayland-logout/blob/master/wayland-logout.py) as well.

**Comment by nwg-piotr** [comment]
Even better!

**Comment by nwg-piotr** [comment]
While at the office, I tested the python version. Works perfectly well.

**Comment by epsilon-0** [comment]
@soreau's based script is dank af.

**Comment by ammen99** [comment]
Fixed in #71

**Comment by epsilon-0** [comment]
thanks @soreau @ammen99
